File: lab8/2-lifetimes/lifetimes.c
#include <stddef.h>
#include <stdio.h>

typedef struct list {
  int val;
  struct list *next;
} list_t;

list_t make_list(int *starting_value) {
  list_t list_node_2 = {*starting_value + 1, NULL};

  list_t list_node_1 = {*starting_value, &list_node_2};

  return list_node_1;
}

int main() {
  int val = 1;
  list_t l = make_list(&val);
  printf("First list element: %d\n", l.val);
  printf("Second list element: %d\n", l.next->val);
}

File: lab8/2-lifetimes/lifetimes.rs
struct List<'a> {
    val: i32,
    next: Option<&'a List<'a>>,
}

#[allow(unused_variables)]
fn main() {
    let my_list = make_list(&1);
}

fn make_list(starting_value: &i32) -> List {
    let list_node_2 = List {
        val: *starting_value + 1,
        next: None,
    };

    let list_node_1 = List {
        val: *starting_value,
        next: Some(&list_node_2),
    };

    return list_node_1;
}
